compiler/rustc_hir/src/hir.rs

```diff
@@ -1847,27 +1840,9 @@ impl Expr<'_> {
 /// Checks if the specified expression is a built-in range literal.
 /// (See: `LoweringContext::lower_expr()`).
 pub fn is_range_literal(expr: &Expr<'_>) -> bool {
-    match expr.kind {
-        // All built-in range literals but `..=` and `..` desugar to `Struct`s.
-        ExprKind::Struct(ref qpath, _, _) => matches!(
-            **qpath,
-            QPath::LangItem(
-                LangItem::Range
-                    | LangItem::RangeTo
-                    | LangItem::RangeFrom
-                    | LangItem::RangeFull
-                    | LangItem::RangeToInclusive,
-                ..
-            )
-        ),
-
-        // `..=` desugars into `::std::ops::RangeInclusive::new(...)`.
-        ExprKind::Call(ref func, _) => {
-            matches!(func.kind, ExprKind::Path(QPath::LangItem(LangItem::RangeInclusiveNew, ..)))
-        }
-
-        _ => false,
-    }
+    // range literals desugar to a struct literal or `RangeInclusive::new(..)`
+    matches!(expr.kind, ExprKind::Call(..) | ExprKind::Struct(..))
+        && expr.span.is_desugaring(DesugaringKind::RangeLiteral)
```

Comment:
I think it should be possible to use comparisons with lang item def ids to detect range expressions without introducing a new `DesugaringKind`.
Using expansion machinery in AST lowering is not a good practice, and I'd like to try eliminating it somehow.

Reply:
Tentatively yes, I just never had time to think what better alternatives are possible.
(Except for `AstPass`, it's used just like it's supposed to be.)

Reply:
Maybe it could work similar to `hir_attrs`.